[SPARK-671] Add initial Kafka Cassandra Stream job.

mpereira opened the pull request.

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+                   )
+
+/**
+  * A config glass used to set the command line arguments.
```
typo for `glass`?

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+                   appName: String = "KafkaRandomFeeder",
+                   brokers: String = "" ,
+                   topics: String = "",
+                   groupId: String = "TODO_USE_A_DIFFERENT_GROUP_ID",
```
Just a heads up, if you use the same hardcoded groupId for all consumers, then you may end up in a situation where consumers read from the queue one a time while all other consumers sit idle. I found that in my scale testing for PR #302

elezar replied:
Yes, I saw that comment. Thanks for the reminder.

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+    StreamingExamples.setStreamingLogLevels()
+
+    val config : Config = Config()
+    if (!getParser("KafkaWordCount").parse(args, config).isDefined) {
```
can use `isEmpty` instead of `!.isDefined`

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+    val sparkBuilder = SparkSession.builder().appName(config.appName)
+
+    val spark = sparkBuilder
+      .master("local[2]")
```
Do we want to make this configurable at all? 

Also no need to save to var `sparkBuilder` only to continue buliding. L168 can be continued ie
```
val spark = SparkSession.builder()
                        .appName(config.appName)
                        .master("local[2]")
                        .getOrCreate()

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+    val wordsPerSecond = config.wordsPerSecond
+    val stream = streamingContext.receiverStream(new RandomWordReceiver(wordsPerSecond))
+
+
```
extra line break?

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+        val producer = new KafkaProducer[String, String](kafkaProperties)
+        p.foreach { r =>
+          println(s"Writing ${r} to Kafka")
+          val msg = new ProducerRecord[String, String](topic, null, r)
```
Personally, I try to stay away from single char variable names. We have to hold so much in our heads as is

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+    val props = new Properties()
+    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokers)
+    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer")
+    props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer")
```
Constants in Scala are [UpperCamel](https://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods) if we're going with convention

elezar replied:
These are Kafka (possibly) Java classes and not in our source.

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+      props.put("sasl.mechanism", "GSSAPI")
+    }
+
+    return props
```
`return` keyword is frowned upon in Scala: https://tpolecat.github.io/2014/05/09/return.html

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+
+    }
+
+    return parser
```
same here

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+
+    val sparkBuilder = SparkSession.builder().appName(config.appName)
+
+    val spark = sparkBuilder
```
can combine these builders

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+      LocationStrategies.PreferConsistent,
+      ConsumerStrategies.Subscribe[String, String](topicsSet, kafkaParams))
+
+    val timestamp = new Date()getTime()
```
this doesn't cause error? `new Date()getTime()` right next to each other?

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+
+    }
+
+    return parser
```
no need for `return`

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,385 @@
+    return parser
+  }
+
+
```
no need for two extra line breaks

elezar commented:
Thanks for the review @samvantran. I have taken your suggestions into account and given it another go.

mpereira commented on tests/jobs/scala/src/main/scala/KafkaCassandra.scala:
```diff
@@ -0,0 +1,257 @@
+      LocationStrategies.PreferConsistent,
+      ConsumerStrategies.Subscribe[String, String](topicsSet, kafkaParams))
+
+    val timestamp = new Date()getTime()
```
`(new Date()).getTime()`?

Also `getTime()` returns a long, do you know if the cassandra driver deals with longs for timestamp columns?

mpereira commented on tests/jobs/scala/src/main/scala/KafkaCassandra.scala:
```diff
@@ -0,0 +1,257 @@
+    val wordCounts = words
+      .map(x => (x, 1L))
+      .reduceByKey(_ + _)
+      .map(x => (timestamp, x._1, x._2))
```
Does this tuple's fields need to be in the same order as the table columns? i.e. word, timestamp, then count?

elezar replied:
I updated the ordering of the columns.

susanxhuynh commented on tests/test_spark.py:
```diff
@@ -249,6 +249,51 @@ def make_credential_secret(envvar, secret_path):
 
+@pytest.mark.sanity
+@pytest.mark.smoke
+def test_kafka_cassandra_stream():
```
Does this test actually involve Kafka? If not, the name is misleading.

mpereira replied:
Not currently, the idea was that it would eventually. Maybe we'll remove this test for now.

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandra.scala:
```diff
@@ -0,0 +1,262 @@
+    val streamingContext = new StreamingContext(spark.sparkContext, Seconds(config.batchSizeSeconds))
+
+    val wordsPerSecond = config.wordsPerSecond
+    val stream = streamingContext.receiverStream(new RandomWordReceiver(wordsPerSecond))
```
since you only use it once, you may as well pass in `config.wordsPerSecond` like you do on L53

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandra.scala:
```diff
@@ -0,0 +1,262 @@
+    //      props.put("sasl.kerberos.service.name", "kafka")
+    //      props.put("security.protocol", "SASL_PLAINTEXT")
+    //      props.put("sasl.mechanism", "GSSAPI")
+    //    }
```
May as well delete if you don't plan to use

samvantran commented on tests/jobs/scala/src/main/scala/KafkaCassandra.scala:
```diff
@@ -0,0 +1,262 @@
+
+
+/** Utility functions for Spark Streaming examples. */
+object StreamingExamples extends Logging {
```
At this point you've got a case class and 3 separate objects in the same file. You may want to start splitting all these out: [as written here](https://docs.scala-lang.org/style/files.html)

mpereira commented:
Hi @samvantran, I think I've addressed your feedback here: https://github.com/mesosphere/spark-build/pull/301/commits/dac227b17c9bbff35d072d604f606d5976679377

elezar commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,137 @@
+  * {PATH_TO_ASSEMBLY} is the location of the snapshot jar containing this application
+  *
+  * Example:
+  * dcos spark run --submit-args='\
```
This should include the other spark options -- like limiting the number of executors.

elezar commented on tests/jobs/scala/src/main/scala/KafkaWordCount.scala:
```diff
@@ -0,0 +1,108 @@
+    val wordCounts = words
+      .map(x => (x, 1L))
+      .reduceByKey(_ + _)
+      .map(x => (new Date().getTime(), x._1, x._2))
```
One question I had was whether this was evaluated only once, or every time we perform the map. Ideally we would want the batch timestamp written to Cassandra. This might be more suited to call to `rdd.saveToCassandra below`.

elezar commented on tests/jobs/scala/src/main/scala/KafkaWordCount.scala:
```diff
@@ -0,0 +1,108 @@
+
+    val spark = StreamingExamples.createSparkSession(config)
+
+    val streamingContext = new StreamingContext(spark.sparkContext, Seconds(config.batchSizeSeconds))
```
We may want to move this closer to where it is being used.

elezar commented on tests/jobs/scala/src/main/scala/KafkaWordCount.scala:
```diff
@@ -0,0 +1,108 @@
+      ConsumerConfig.GROUP_ID_CONFIG -> groupId,
+      ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG -> classOf[StringDeserializer],
+      ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG -> classOf[StringDeserializer]
+    ) ++ (
```
Scala. Gotta love it!

mpereira replied:
As a lisp weenie, infix operators make me sad.

elezar commented on tests/jobs/scala/src/main/scala/RandomWordReceiver.scala:
```diff
@@ -0,0 +1,45 @@
+
+class RandomWordReceiver(wordsPerSecond: Float) extends Receiver[String](StorageLevel.MEMORY_ONLY_2) with Logging {
+  val waitFor: Long = (1000L / wordsPerSecond).toLong
+  val inputList: Array[String] = "the quick brown fox jumps over the lazy dog".split(" ")
```
Note that this currently has a pretty limited set of words, but this is probably OK for testing.

elezar commented on tests/jobs/scala/src/main/scala/StreamingExamples.scala:
```diff
@@ -0,0 +1,90 @@
+)
+
+/** Utility functions for Spark Streaming examples. */
+object StreamingExamples extends Logging {
```
This `object` name could be renamed a bit?

elezar replied:
(I'm happy to leave it as is for this PR though)

elezar commented on tests/test_spark.py:
```diff
@@ -249,6 +249,52 @@ def make_credential_secret(envvar, secret_path):
+@pytest.mark.sanity
+@pytest.mark.smoke
+@pytest.mark.skip("TODO")
+def test_kafka_cassandra_stream():
```
Let's rename this test to `test_cassandra_stream` and remove the `sanity` mark.

elezar commented on tests/jobs/scala/src/main/scala/KafkaCassandraStream.scala:
```diff
@@ -0,0 +1,137 @@
+  * http://infinity-artifacts.s3.amazonaws.com/autodelete7d/dcos-spark-scala-tests-assembly-0.1-SNAPSHOT.jar \
+  * --numberOfRecords 100000 --numberOfPartitions 10 --minColumnsLength 20 --maxColumnsLength 30'
+  */
+object KafkaCassandraStream {
```
This class is useful for testing writing to Cassandra (the original intent of SPARK-671), but does not interact with Kafka. My proposal would be to drop it (and the associated integration test) from this PR and create a new PR that only addresses the Spark->Cassandra interaction. This was definitely my bad for hi-jacking the PR to add the Kafka->Spark->Cassandra wordcount example. Sorry.

mpereira commented on tests/jobs/scala/src/main/scala/StreamingExamples.scala:
```diff
@@ -0,0 +1,90 @@
+          c.copy(wordsPerSecond = x.toFloat)).text("The rate at which words should be written to Kafka")
+      } else {
+        opt[Unit]("shouldWriteToCassandra").action((_, c) =>
+          c.copy(shouldWriteToCassandra = true)).text("Write word counts to Cassandra")
```
Wouldn't `--shouldWriteToCassandra false` disable it?

mpereira commented:
@elezar I think I addressed everything here: https://github.com/mesosphere/spark-build/pull/301/commits/f039fe296002ea149231fe31d26729f5bef6b82a

mpereira commented:
This is how I've been testing it manually: 

### Submit producer job

```
dcos spark run --name=kafka-cassandra-streaming-spark --submit-args='
--conf spark.cores.max=5
--conf spark.mesos.executor.docker.image=mesosphere/spark:latest
--conf spark.mesos.executor.home=/opt/spark/dist
--class KafkaRandomFeeder http://infinity-artifacts.s3.amazonaws.com/autodelete7d/dcos-spark-scala-tests-assembly-kafka-cassandra-streaming.jar
--brokers kafka-0-broker.kafka-cassandra-streaming-kafka.autoip.dcos.thisdcos.directory:1025,kafka-1-broker.kafka-cassandra-streaming-kafka.autoip.dcos.thisdcos.directory:1025,kafka-2-broker.kafka-cassandra-streaming-kafka.autoip.dcos.thisdcos.directory:1025
--topics randomwords
--wordsPerSecond 100
--appName KafkaRandomFeeder
'
```

### Submit consumer job

```
dcos spark run --name=kafka-cassandra-streaming-spark  --submit-args='
--conf spark.cores.max=5
--conf spark.mesos.executor.docker.image=mesosphere/spark:latest
--conf spark.mesos.executor.home=/opt/spark/dist
--conf spark.cassandra.connection.host=node-0-server.kafka-cassandra-streaming-cassandra.autoip.dcos.thisdcos.directory
--conf spark.cassandra.connection.port=9042
--class KafkaWordCount http://infinity-artifacts.s3.amazonaws.com/autodelete7d/dcos-spark-scala-tests-assembly-kafka-cassandra-streaming.jar
--brokers kafka-0-broker.kafka-cassandra-streaming-kafka.autoip.dcos.thisdcos.directory:1025,kafka-1-broker.kafka-cassandra-streaming-kafka.autoip.dcos.thisdcos.directory:1025,kafka-2-broker.kafka-cassandra-streaming-kafka.autoip.dcos.thisdcos.directory:1025
--topics randomwords
--batchSizeSeconds 10
--appName KafkaWordCount
'
```

### Check Kafka topic consumers

```
dcos task exec -it kafka-0-broker__d2ec3276-6805-448a-9c89-adf54ffaaccf bash -c 'export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/jre/) && export KAFKA_BIN_DIRECTORY=$(ls -d $MESOS_SANDBOX/kafka_*/bin) && $KAFKA_BIN_DIRECTORY/kafka-consumer-groups.sh --bootstrap-server kafka-0-broker.kafka-cassandra-streaming-kafka.autoip.dcos.thisdcos.directory:1025 --group kafka_cassandra_stream --describe'
```

### Check Cassandra table state

```
dcos task exec -it node-0-server bash -c 'export CQLSH=$(ls -d ./apache-cassandra-*/bin/cqlsh) && $CQLSH node-0-server.kafka-cassandra-streaming-cassandra.autoip.dcos.thisdcos.directory 9042 -e "select * from kafka_cassandra_stream.records;"'
```
